### DISCUSSION TITLE
esphome update TFT from 4.2.5 to 4.2.6 with ble tracker working, or?

### DISCUSSION THREAD

**Discussion by MichaelHeimann** [Blackymas/NSPanel_HA_Blueprint#1823] [General]
I wanted to test the improvements with esphome 2024.2.0 in regards to tft upload, especially since I do use BLE tracker which in the past caused crashes due to low heap. (actually heap was decreasing fast during upload and crashed when it got too low).

Now, I've just updated to 4.2.5 with esphome 2023.12.9 (? not 100% sure, latest greatest release before 2024.2.0) to 4.2.6 with esphome 2024.2.0. I do have esp32_ble_tracker and addon_climate in my config and still, the update went through and ha now shows 4.2.6 for tft, firmware and blueprint.

Only thing that keeps me worried is that the upload startet from 86.53 %. Is there some magic going on where it sees if the tft files are the same up to a certain point? 
The free heap kept decreasing this time, also, but the upload did finish so I think the issue is not solved but mitigated enough.

Only question I have: How can I check if my tft is valid now? Do I have a mixed 4.2.5 and 4.2.6 version? running?

Here's the log:

[logs_nsbuero_logs.txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14370656/logs_nsbuero_logs.txt)

**Comment by edwardtfn** [comment]
> Only thing that keeps me worried is that the upload startet from 86.53 %. Is there some magic going on where it sees if the tft files are the same up to a certain point?

That's fine. The latest Nextion protocol allows resuming with the parts of the file that has changed. 😉

> Only question I have: How can I check if my tft is valid now? Do I have a mixed 4.2.5 and 4.2.6 version? running?

The system will work and won't report version mismatch, as only minor version differemce, but I  highly recommend you update the TFT to the same version to prevent issues.
To update that, you now can just select the panel's model on the dropdown componente in your device's page and then click update, it will get the latest from GitHub directly.
If for some reason you prefer (or have) to use your local hosted copy, please make sure to update your server with v4,2,6.

**Reply by MichaelHeimann** [reply]
Resuming with changed parts is what I assumed. But where is it detected, that the first 86% are the same and thus resuming is possible? Did the esp verify that? I'd assume it needs to binary compare the two files and since it doesn't have them both... 
Just being curious.

Yeah, my panels are in a IOT vlan without internet, so I need a local source for the tft. But I prefer that anyways, because it removes WAN out of the serial upload error equation. 

But I'll update them all to be on 4.2.6 everywhere.
To me it is a great improvement if bt and tft upload work together. I'm wondering if it also works when the tft needs a complete upload and not only the last few percent.

**Comment by edwardtfn** [comment]
By the way, could you please share your panel's yaml? I'd like to duplicate this memory thing here.
Please make sure you have removed any private info like wifi credentials before sharing.

**Reply by MichaelHeimann** [reply]
sure, here it is
[nsbuero.yaml.txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14383014/nsbuero.yaml.txt)

**Comment by MichaelHeimann** [comment]
OK, I've now tried to update the other nspanels and all but 2 went through. these two repeatedly reported heap going below 5k which seems to be too low and I experience log connection drops, restarts or plain tft upload failures.
I've flashed these two by going the no-ble firmware -> tft upload -> ble firmware workaround.

Note: when using ble in the firmware, the free heap is going down by up to 8k during upload (and it's only 14% of the tft this time):
e.g. from log with ble: 
```
[08:35:04][D][nextion.upload.idf:496]:   URL: http://10.250.23.6:8123/local/nspanel_eu.tft
[08:35:04][D][nextion.upload.idf:497]:   File size: 7351628 bytes
[08:35:04][D][nextion.upload.idf:498]:   Free heap: 23160
[08:35:09][D][nextion.upload.idf:238]: Uploaded 0.06 %, remaining 7347532 bytes, free heap: 25456 bytes
[08:35:10][D][nextion.upload.idf:238]: Uploaded 86.53 %, remaining 990540 bytes, free heap: 18672 bytes
[08:35:31][D][nextion.upload.idf:238]: Uploaded 99.40 %, remaining 44364 bytes, free heap: 10752 bytes
[08:35:32][D][nextion.upload.idf:238]: Uploaded 100.00 %, remaining 0 bytes, free heap: 19920 bytes
```

without ble it's not going down:
```
[10:09:44][D][nextion.upload.idf:496]:   URL: http://10.250.23.6:8123/local/nspanel_eu.tft
[10:09:44][D][nextion.upload.idf:497]:   File size: 7351628 bytes
[10:09:44][D][nextion.upload.idf:498]:   Free heap: 142108
[10:09:49][D][nextion.upload.idf:238]: Uploaded 0.06 %, remaining 7347532 bytes, free heap: 139772 bytes
[10:09:49][D][nextion.upload.idf:238]: Uploaded 86.53 %, remaining 990540 bytes, free heap: 132736 bytes
... doesn't drop below 132000 bytes
[10:10:10][D][nextion.upload.idf:238]: Uploaded 99.95 %, remaining 3404 bytes, free heap: 134428 bytes
[10:10:10][D][nextion.upload.idf:238]: Uploaded 100.00 %, remaining 0 bytes, free heap: 140188 bytes
```

so ble is causing ram usage to "jitter". If we can stop that, ble + tft upload would probably work reliably.

**Reply by edwardtfn** [reply]
You can try adding this to your panel's yaml, as a customization, so it will stop the ble scan when you start the upload. If the issue persists, we can try something to disable the clients also:

```yaml
esp32_ble_tracker:
  id: ble_tracker

script:
  - id: !extend upload_tft
    then:
      - lambda: |-
          static const char *const TAG = "CUSTOM.script.upload_tft";
          ble_tracker->dump_config();
          ESP_LOGI(TAG, "Stopping BLE Tracker scan...");
          ble_tracker->stop_scan();
          ESP_LOGI(TAG, "Disabling BLE Tracker scan...");
          ble_tracker->set_scan_active(false);
```

**Comment by edwardtfn** [comment]
We found a memory leak that may be connected to the issue you reported:

- #1832

That is already fixed in dev and will be part of the next release.
